installinator/src/reporter.rs

```diff
@@ -49,129 +78,418 @@ impl ProgressReporter {
             report_backend,
         };
         (ret, event_sender)
     }
 
-    pub(crate) fn start(mut self) -> tokio::task::JoinHandle<()> {
+    /// Starts the event reporter in a new Tokio task.
+    ///
+    /// The returned task handle exits once at least one peer has accepted a
+    /// terminal event report.
+    pub(crate) fn start(mut self) -> JoinHandle<()> {
         tokio::spawn(async move {
             // TODO: jitter?
-            let mut interval = time::interval(Duration::from_secs(1));
-            interval.set_missed_tick_behavior(time::MissedTickBehavior::Delay);
-            interval.tick().await;
-
+            let mut report_interval = time::interval(Duration::from_secs(2));
+            report_interval
+                .set_missed_tick_behavior(time::MissedTickBehavior::Delay);
+            // Do not tick report_interval immediately. We'd like to generate a
+            // report at the start of the loop to ensure that report_tx is
+            // initialized right away and that new peers receive reports ASAP.
+
+            let (peers_tx, peers_rx) = watch::channel(None);
+            let (discover_cancel_tx, discover_cancel_rx) =
+                coop_cancel::new_pair();
+            let (report_tx, report_rx) = watch::channel(None);
+            let (completion_tx, mut completion_rx) = mpsc::channel(1);
+            self.discover_task =
+                Some(self.spawn_discover_task(peers_tx, discover_cancel_rx));
+            self.reconcile_peers_task =
+                Some(self.spawn_reconcile_peers_task(peers_rx, report_rx));
+
+            let mut reports_done = false;
             let mut events_done = false;
 
+            // The goal of this loop is to:
+            //
+            // * process all events
+            // * get a fully completed report out to at least one peer which
+            //   accepts this report
+            //
+            // The state machine is implemented via a `tokio::select` loop. The
+            // general logic is:
+            //
+            // * on receiving an event from the update engine
+            //   - add event to self.buffer
+            //   - if event receiver is closed, the update engine has completed, so
+            //     mark events_done = true, and indicate that the next report should
+            //     be sent right away
+            //
+            // * on report_interval tick:
+            //   - generate report
+            //   - if events_done is true, mark report as complete and pass in a
+            //     completion channel sender
+            //   - send message to report tasks
+            //
+            // * on receiving completion channel message (indicating that at
+            //   least one completed report has been received by a peer):
+            //   - mark reports_done = true
+            //
+            // If events_done and reports_done are true, then the loop's goal is
+            // met, so exit the loop.
             loop {
```

Comment:
I'm tempted to suggest

```suggestion
            while !(events_done && reports_done) {
```

so that the body of the loop is _just_ the `select!` (moving the discovery cancellation below the loop entirely). Entirely subjective though; up to you.